### PR TITLE
Tdl 22064 add integration tests

### PR DESCRIPTION
**Pull request by Vi6hal**
# Description of change
[TDL-22330](https://jira.talendforge.org/browse/TDL-22330) Fix ` jobs` stream records with same id's are being replicated with different "workflow_id" (parent stream) multiple times
Includes Following Tests
- All Fields Test
- Automatic Fields Test
- Start Date Test
- Bookmark Test
- Interruptible Sync Test

Fix CCi Configuration

# Manual QA steps
 -  Ran and Verified all Tests Locally
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_circle_ci/streams/jobs.py [hunk 1]
```diff
@@ -69,6 +70,12 @@ def sync(self, state: Dict, schema: Dict, stream_metadata: Dict, transformer: Tr
                 for index, (workflow_id, pipeline_id) in enumerate(pipelines[start_index:], max(start_index, 1)):
                     LOGGER.info("Syncing jobs for workflow *****%s (%s/%s)", workflow_id[-4:], index, prod_len)
                     for rec in self.get_records(workflow_id):
+
+                        # The same job can be fetched multiple times as it relates to 
+                        # multiple parents i.e workflows, this creates a composite key with the parent_id
+                        comp_key = workflow_id + rec["id"]
+                        hash_encoded = comp_key.encode('utf-8')
+                        rec['_sdc_record_hash'] = sha256_encode(hash_encoded).hexdigest()
```
Comment (leslievandemark) [inline tap_circle_ci/streams/jobs.py:78]: Is there a ticket for this change? could you add it to the description

File: tests/test_interrupted_sync.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                            msg="For pending sync streams - {}, second sync record count \
+                                                should be more than or equal to first sync".format(
+                                stream
+                            ),
+                        )
+                    else:
+                        raise Exception(f"Invalid state of stream {stream} in interrupted state")
+
+                elif expected_replication_method == self.FULL_TABLE:
+
```
Comment (bhtowles) [inline tests/test_interrupted_sync.py:168]: Additional FULL_TABLE verifications:
for rec in interrupted_records:
                        self.assertIn(rec, full_records, msg='full table record in interrupted sync not found in full sync')

                # Verify at least 1 record was replicated for each stream
                self.assertGreater(interrupted_record_count, 0)
                print(f"{stream} resumed sync records replicated: {interrupted_record_count}")

File: tests/test_start_date.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                # Expected values
+                expected_primary_keys = self.expected_primary_keys()[stream]
+                expected_metadata = self.expected_metadata()[stream]
+                expected_start_date_1 = self.timedelta_formatted(self.start_date_1, self.START_DATE_FORMAT, -1)
+                expected_start_date_2 = self.timedelta_formatted(self.start_date_2, self.START_DATE_FORMAT, -1)
+
+                # Collect information for assertions from syncs 1 & 2 base on expected values
+                record_count_sync_1 = record_count_by_stream_1.get(stream, 0)
+                record_count_sync_2 = record_count_by_stream_2.get(stream, 0)
+
```
Comment (bhtowles) [inline tests/test_start_date.py:88]: self.assertIn(stream, SYNCED_MESSAGES_BY_STREAM_1.keys(),
                              msg=f"No records for stream: {stream} in sync 1")

File: tests/test_start_date.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                primary_keys_list_2 = [
+                    tuple(message.get("data").get(expected_pk) for expected_pk in expected_primary_keys)
+                    for message in synced_records_2.get(stream, {}).get("messages", [])
+                    if message.get("action") == "upsert"
+                ]
+
+                primary_keys_sync_1 = set(primary_keys_list_1)
+                primary_keys_sync_2 = set(primary_keys_list_2)
+
+                if expected_metadata.get(self.OBEYS_START_DATE):
```
Comment (bhtowles) [inline tests/test_start_date.py:103]: shouldn't this be:
 if expected_metadata[stream].get(self.OBEYS_START_DATE):

**Review by bhtowles — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,22 +1,46 @@
...
+      - run:
+          name: 'JSON Validator'
+          command: |
+            source /usr/local/share/virtualenvs/tap-tester/bin/activate
+            stitch-validate-json tap_circle_ci/schemas/*.json
       - run:
           name: 'pylint'
           command: |
             source /usr/local/share/virtualenvs/tap-circle-ci/bin/activate
-            make test_alpha
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply .circleci/config.yml:19]: This tap has a makefile created with few commands, this `test_alpha` runs the pylint command with a elongated ignore list which is not needed anymore, hence this has been removed

this is the command that it runs:
```makefile
test_alpha:
	pylint tap_circle_ci -d broad-except,chained-comparison,empty-docstring,\
	fixme,invalid-name,line-too-long,missing-class-docstring,\
	missing-function-docstring,missing-module-docstring,no-else-raise,\
	no-else-return,too-few-public-methods,too-many-arguments,\
	too-many-branches,too-many-lines,too-many-locals,ungrouped-imports,\
	wrong-spelling-in-comment,wrong-spelling-in-docstring,\
	unnecessary-comprehension,too-many-nested-blocks,redefined-builtin,\
	bad-whitespace,dangerous-default-value,bad-continuation,unused-variable,\
	logging-format-interpolation,unused-argument
```

[Makefile](https://github.com/singer-io/tap-circle-ci/blob/f1f3fb308aa00407c2510722ae562b5d3a0076a7/Makefile#L9)

File: tests/test_bookmark.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
+from base import CircleCiBaseTest
+from datetime import datetime as dt
+from datetime import timedelta
+
+import dateutil.parser
+from tap_tester import connections, menagerie, runner
+from tap_tester.logger import LOGGER
+from singer.utils import strptime_to_utc
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_bookmark.py]: Yes, you are right! i have updated this please check.

File: tests/test_interrupted_sync.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
+from base import CircleCiBaseTest
+from singer.utils import strptime_to_utc
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_interrupted_sync.py]: fixed

File: tests/test_all_fields.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                # Verify that more than just the automatic fields are replicated for each stream.
+                self.assertTrue(
+                    expected_automatic_keys.issubset(expected_all_keys),
+                    msg=f'{expected_automatic_keys - expected_all_keys} is not in "expected_all_keys"',
+                )
+
+                messages = synced_records.get(stream)
+                # Collect actual values
+                actual_all_keys = set()
+                for message in messages["messages"]:
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_all_fields.py:74]: added assertion to check if the stream has any synced records before collecting records

File: tests/test_all_fields.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        conn_id = connections.ensure_connection(self)
+
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Table and field selection
+        test_catalogs_all_fields = [
+            catalog for catalog in found_catalogs if catalog.get("tap_stream_id") in expected_streams
+        ]
+
+        self.perform_and_verify_table_and_field_selection(conn_id, test_catalogs_all_fields)
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_all_fields.py:37]: the `select_all_fields` is true by default in base.py, hence not added here

File: tests/test_automatic_fields.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        # Run initial sync
+        self.run_and_verify_sync(conn_id)
+        synced_records = runner.get_records_from_target_output()
+
+        for stream in expected_streams:
+            with self.subTest(stream=stream):
+                # Expected values
+                expected_keys = self.expected_automatic_fields().get(stream)
+                expected_primary_keys = self.expected_primary_keys()[stream]
+
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_automatic_fields.py:52]: added assertion to check if the stream has any synced records before collecting record

File: tests/test_start_date.py (hunk 3 above)
Comment (Vi6hal) [reply tests/test_start_date.py:88]: updated the assertions

File: tests/test_start_date.py (hunk 4 above)
Comment (Vi6hal) [reply tests/test_start_date.py:103]:
```python
            with self.subTest(stream=stream):
                # Expected values
                expected_primary_keys = self.expected_primary_keys()[stream]
                expected_metadata = self.expected_metadata()[stream]
```

expected metadata is already fetched for the current stream, hence wont be needed to use the stream name again

File: tap_circle_ci/streams/jobs.py (hunk 1 above)
Comment (Vi6hal) [reply tap_circle_ci/streams/jobs.py:78]: [TDL-22330](https://jira.talendforge.org/browse/TDL-22330) this is the ticket for this change,  i have updated the P.R description.

File: tests/test_bookmark.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+                if expected_replication_method == self.INCREMENTAL:
+                    # Collect information specific to incremental streams from syncs 1 & 2
+                    replication_key = next(iter(expected_replication_keys[stream]))
+                    first_bmk_value = {k: strptime_to_utc(v) for k, v in first_bookmark.items()}
+                    second_bmk_value = {k: strptime_to_utc(v) for k, v in second_bookmark.items()}
+                    simulated_bmk_value = {key: strptime_to_utc(value) for key, value in new_state["bookmarks"][stream].items()}
+
+                    # Verify the first sync sets a bookmark of the expected form
+                    self.assertIsNotNone(first_bookmark)
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_bookmark.py:121]: I have added these assertions

File: tests/test_bookmark.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+                            LOGGER.info("Key not found in the second bookmark value %s %s %s",replication_key_value, simulated_bookmark,second_bookmark_value_utc_value)
+
+                elif expected_replication_method == self.FULL_TABLE:
+
+                    # Verify the syncs do not create bookmark for full table streams
+                    self.assertEqual(first_bookmark,{})
+                    self.assertEqual(second_bookmark,{})
+
+                    # Verify the number of records in the second sync is the same as the first
+                    self.assertEqual(second_sync_count, first_sync_count)
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_bookmark.py:224]: I have added this assertions

File: tests/test_bookmark.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+                            # check that all records synced hold a replication key value higher than the simulated bookmark value
+                            self.assertGreaterEqual(
+                                replication_key_value,
+                                simulated_bookmark,
+                                msg="Second sync records do not repeat the previous bookmark.",
+                            )
+
+                        except KeyError:
+                            LOGGER.info("Key not found in the second bookmark value %s %s %s",replication_key_value, simulated_bookmark,second_bookmark_value_utc_value)
+
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_bookmark.py:212]: I have added this assertion

File: tests/test_bookmark.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+                if expected_replication_method == self.INCREMENTAL:
+                    # Collect information specific to incremental streams from syncs 1 & 2
+                    replication_key = next(iter(expected_replication_keys[stream]))
+                    first_bmk_value = {k: strptime_to_utc(v) for k, v in first_bookmark.items()}
+                    second_bmk_value = {k: strptime_to_utc(v) for k, v in second_bookmark.items()}
+                    simulated_bmk_value = {key: strptime_to_utc(value) for key, value in new_state["bookmarks"][stream].items()}
+
+                    # Verify the first sync sets a bookmark of the expected form
+                    self.assertIsNotNone(first_bookmark)
+
```
[the remark this answers is not in the document]
Comment (Vi6hal) [reply tests/test_bookmark.py:122]: I have added this assertion

**Review by bhtowles — APPROVED** [review]

File: tap_circle_ci/streams/jobs.py (hunk 1 above)
Comment (Vi6hal) [reply tap_circle_ci/streams/jobs.py]: I have removed the `_sdc_record_hash` as using a composite key will solve the issue of record duplication, these are the updated primary keys for jobs stream
Previous:
```python
    key_properties = ["id","_sdc_record_hash"]
```
Updated:
```python
    key_properties = ["id","_workflow_id"]
```

**Pull request merged**